### ISSUE TITLE
训练首个神经网络，网上教程的demo练习

### ISSUE THREAD

**Issue by wubaigit** [wubaigit/wubai-blog#9]
> 资料来源：[训练首个神经网络：基本分类](https://www.tensorflow.org/tutorials/keras/basic_classification?hl=zh-cn)

**首先，导入 Tensorflow， tf.keras 和 依赖**

```python
import tensorflow as tf
from tensorflow import keras

import numpy as np
import matplotlib.pyplot as plt
```

打印出 tensorflow 的版本

```python
print(tf.__version__)
```

**接着，我们要导入 fashion mnist 的数据集。**现在可以使用 tensorflow 直接获取，首次运行会自动去网上下载该数据集并保存在本地

```python
fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
```

加载数据集会返回 4 个 Numpy 数组，前两个分别是【用于训练的图像】和【用于训练的标签】，后两个分别是【用于测试的图像】和【用于测试的标签】，即，train_images 和 train_labels 数组是训练集，test_images 和 test_labels 数组是测试集

图像（image）为 28x28 的 Numpy 数组，像素值介于 0 和 255 之间，标签（label）是整数数组，介于 0 到 9 之间，分别对应一种服饰的类别

为了方便，我们使用一个数组去维护 数字 和 服饰类别 的对应表

```python
class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat', 'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']
```

训练集一共 60000 张图片，每张图片都表示为 28x28 像素——

```python
print(train_images.shape) # (60000, 28, 28)
```

相对应具有 60000 个标签——

```python
print(len(train_labels))  # 60000
print(train_labels) # array([9,0,0,...,3,0,5],dtype=uint8)
```

测试集同理，只是，测试集是 10000 张图片

**做预处理**

```python
plt.figure()
plt.imshow(train_images[0])
plt.colorbar()
plt.grid(False)

train_images = train_images / 255.0
test_images = test_images / 255.0
```

`pyplot` 是一组命令样式函数，使 `matplotilib` 像 `MATLAB` 一样工作，

每个 `pyplot` 函数都会对图形进行一些更改，例如，_创建图形_，_在图形中创建绘图区域_，_在绘图区域中绘制一些线条_，*使用标签装饰图形*等

```python
plt.figure(figsize=(10, 10))
for i in range(25):
  plt.subplot(5, 5, i+1)
  plt.xticks([])
  plt.yticks([])
  plt.grid(False)
  plt.imshow(train_images[i], cmap=plt.cm.binary)
  plt.xlabel(class_names[train_labels[i]])
```

**构建模型**

构建神经网络需要先配置模型的层，然后再编译模型

**设置层**

神经网络的基本构造块是层。层从馈送到其中的数据中提取表示结果。大部分深度学习都会把简单的层连在一起

```python
model = keras.Sequential([
  keras.layers.Flatten(input_shape=(28, 28)),
  keras.layers.Dense(128, activation=tf.nn.relu),
  keras.layers.Dense(10, activation=tf.nn.softmax)
])
```

keras.Sequential 是顺序模型 API，第一层 tf.keras.layers.Flatten 将图像格式从而为数组转化为一维数组，只改动数据的格式；扁平化像素后，该网络包含了两个 tf.keras.layers.Dense 层序列，这些层是密集连接或全连接神经层；第一层 Dense 层具有 128 个节点（或神经元）；第二层 是具有 10 个节点的 softmax 层。该层会返回一个具有 10 个概率得分的数组（这些得分的总和为 1。每个节点包含一个得分，表示当前图像属于 10 个类别中某一个的概率）

**编译模型**

模型还需要再进行几项设置才可以开始训练，这些设置会添加到模型的编译步骤

1. 损失函数：衡量模型在训练期间的准确率。
2. 优化器：根据模型看到的数据及其损失函数更新模型的方式。
3. 指标：用于监控训练和测试步骤。

```python
model.compile(
  optimizer=tf.train.AdamOptimizer(),
  loss='sparse_categorical_crossentropy',
  metrics=['accuracy'], # 使用准确率作为指标
)
```

**训练模型**

1. 将训练数据馈送到模型中，本例中为 train_images 和 train_labels 数组
2. 模型学习将图像与标签相关联
3. 要求模型对测试集进行预测

```python
# 调用 model.fit ，使模型与训练数据“拟合”
model.fit(train_images, train_labels, epochs=5)

# 按 batch 计算在某些输入数据上模型的误差
test_loss, test_acc = model.evaluate(test_images, test_labels)
print('准确率：', test_acc)
```

如果机器学习模型在新数据上的表现不如在训练数据上的表现，就表示出现过拟合

**对一些图像进行预测**

```python
predictions = model.predict(test_images)  # 返回一个 numpy array

def plot_image(i, predictions_array, true_label, img):
  predictions_array, true_label, img = predictions_array[i], true_label[i], img[i]
  plt.grid(False)
  plt.xticks([])
  plt.yticks([])

  plt.imshow(img, cmap=plt.cm.binary)

  predicted_label = np.argmax(predictions_array)
  if predicted_label == true_label:
    color = 'blue'
  else:
    color = 'red'

  plt.xlabel(
    "{} {:2.0f}% ({})".format(class_names[predicted_label],
    100*np.max(predictions_array),
    class_names[true_label]),
    color=color
  )

def plot_value_array(i, predictions_array, true_label):
  predictions_array, true_label = predictions_array[i], true_label[i]
  plt.grid(False)
  plt.xticks([])
  plt.yticks([])
  thisplot = plt.bar(range(10), predictions_array, color="#777777")
  plt.ylim([0, 1])
  predicted_label = np.argmax(predictions_array)

  thisplot[predicted_label].set_color('red')
  thisplot[true_label].set_color('blue')

num_rows = 5
num_cols = 3
num_images = num_rows*num_cols
plt.figure(figsize=(2*2*num_cols, 2*num_rows))
for i in range(num_images):
  plt.subplot(num_rows, 2*num_cols, 2*i+1)
  plot_image(i, predictions, test_labels, test_images)
  plt.subplot(num_rows, 2*num_cols, 2*i+2)
  plot_value_array(i, predictions, test_labels)

plt.show()
```

总结，训练一个神经网络至少要经过——

- 导入数据集
- 预处理数据
- 构建模型
  - 配置模型
  - 编译模型
  - 训练模型
  - 评估准确率
- 对新图像做出预测
